packages/hooks/src/test/index.test.js

```diff
@@ -943,3 +945,86 @@ test( 'checking hasFilter with named callbacks and removeAllActions', () => {
 	expect( hasFilter( 'test.filter', 'my_callback' ) ).toBe( false );
 	expect( hasFilter( 'test.filter', 'my_second_callback' ) ).toBe( false );
 } );
+
+describe( 'async filter', () => {
+	test( 'runs all registered handlers', async () => {
+		addFilter( 'test.async.filter', 'callback_plus1', ( value ) => {
+			return new Promise( ( r ) =>
+				setTimeout( () => r( value + 1 ), 10 )
+			);
+		} );
+		addFilter( 'test.async.filter', 'callback_times2', ( value ) => {
+			return new Promise( ( r ) =>
+				setTimeout( () => r( value * 2 ), 10 )
+			);
+		} );
+
+		expect( await applyAsyncFilters( 'test.async.filter', 2 ) ).toBe( 6 );
+	} );
+
+	test( 'aborts when handler throws an error', async () => {
+		const sqrt = jest.fn( async ( value ) => {
+			if ( value < 0 ) {
+				throw new Error( 'cannot pass negative value to sqrt' );
+			}
+			return Math.sqrt( value );
+		} );
+
+		const plus1 = jest.fn( async ( value ) => {
+			return value + 1;
+		} );
+
+		addFilter( 'test.async.filter', 'callback_sqrt', sqrt );
+		addFilter( 'test.async.filter', 'callback_plus1', plus1 );
+
+		await expect(
+			applyAsyncFilters( 'test.async.filter', -1 )
+		).rejects.toThrow( 'cannot pass negative value to sqrt' );
+		expect( sqrt ).toHaveBeenCalledTimes( 1 );
+		expect( plus1 ).not.toHaveBeenCalled();
+	} );
+} );
+
+describe( 'async action', () => {
+	test( 'runs all registered handlers sequentially', async () => {
+		const outputs = [];
+		const action1 = async () => {
+			outputs.push( 1 );
+			await new Promise( ( r ) => setTimeout( () => r(), 10 ) );
+			outputs.push( 2 );
+		};
+
+		const action2 = async () => {
+			outputs.push( 3 );
+			await new Promise( ( r ) => setTimeout( () => r(), 10 ) );
+			outputs.push( 4 );
+		};
+
+		addAction( 'test.async.action', 'action1', action1 );
+		addAction( 'test.async.action', 'action2', action2 );
+
+		await doAsyncAction( 'test.async.action' );
+		expect( outputs ).toEqual( [ 1, 2, 3, 4 ] );
```

Comment:
Actually, I wonder if this should be the right order. I feel like the correct order should be `1, 3, 2, 4` (in other words, internally we'd use `Promise.all` to allow for parallel actions).

Reply:
`addAction` also has `priority` as argument, both in JS and PHP versions. So the order matters and can be influenced.

The only difference between actions and filters is that filters process a "value" that is passed from one handler to another.

Reply:
Yes true but I feel that for most actions, we don't really care about the order. I wonder if it should be an option. For instance, this has the waterfall effect if multiple plugins trigger REST APIs for instance in an async action...

Anyway, I can leave with both approaches.

Reply:
Now I understand, the registered actions could also run in parallel, and there would be one `Promise.all` or `Promise.allSettled` that waits and only then `doAsyncAction` returns.

Why not, but that means entering a new territory where there is not precedent with existing PHP and JS actions. These all run sequentially.

Do you have a use case for async actions? Filters are easy, they always must be sequential, but actions can run in different modes.

Reply:
think about the "post save" action we have or the "pre save" action. How do you see plugins using this one? should they be parallel or sequential actions?

My use-case is "registration of data views actions". Like maybe plugins have some async dependencies before registering their actions...

Reply:
I agree that it's bit of a new thing, but async actions are inherently different than php actions anyway. I'm very ok starting with sequential actions though.

Reply:
Theoretically, order should matter for both actions and filters, specifically if we want to mirror the hook behavior on the PHP side. Order does matter there, and is predictable. It's important for consistent behavior when you have multiple functions hooked to the same action. 

Imagine you're hooking 10 functions from separate plugins, each of them registering tabs in a tabbed layout. If you care about the tab order (and you likely do), then you'll likely care about the consistency of the action order.

Reply:
> think about the "post save" action we have or the "pre save" action. How do you see plugins using this one?

The known use cases for `preSavePost` are:
- validation. E.g., the ACF plugin does its own form validation and aborts the save early if it fails: https://github.com/WordPress/gutenberg/pull/44971#issuecomment-1283895553
- modifying the `edits`, e.g., to calculate some derived fields from modified ones. Block hooks could use this, but it's not yet an use case that's been validated by actually implementing it.

Validations can run in parallel, because the result is a boolean success/fail. One failed validation is sufficient to fail them all.

Modifying edits must run sequentially, the output of one handler is fed to the next one.

For `savePost` (after the entity was saved to REST) the known use case is saving metaboxes. After the post itself is saved, the metabox forms are saved to `_wpMetaBoxUrl`. These handlers could run in parallel.

After some consideration, I believe that async actions are inherently parallel, because the input of any action doesn't depend on the output of any other. While async filters are inherently sequential.

I'm not sure about handling errors. Until now the sync hook didn't really support throwing errors. Stale data would be left inside `hooksStore.__current` if any handler threw an exception, there was no `try/catch` block.

In the current state of this PR, throwing an error immediately aborts the hook run. Further handlers are called only when all previous handlers succeed. It's not possible, for example, that a later handler recovers from an error thrown by an earlier one. Like you can do with promises: `promise.then().catch().then()`. I'm not sure if this lack of flexibility is a problem or not.

In PHP, a handler can return a `wp_error` and the next handler has complete freedom what to do with that error.

Reply:
I'm still hesitant about which one do we need more parallel or not but I'm thinking that we should pick one for now and see how far we can get with it.

Your call, if you think "serial/waterfall" is better as default, we can go with it for now and consider a separate function later for parallel if need be.

Reply:
Looking at the base webpack codebase, they have 42 async hook, and only 6 of them are parallel. The majority, 36, are serial. That suggests that parallel actions are an exception.

So, yes, I'll go with `doActionAsync` and `applyFiltersAsync` being serial by default. We can add `doActionAsyncParallel` at any time later when there is a demand.

Also, `doActionAsync` seems to be a better name than `doAsyncAction`. It follows a convention where the base name is `doAction` and there is a series of modifiers. Core has `do_action_deprecated` or `do_action_ref_array`, so this convention is already established. I'll modify the PR accordingly.